pilot/pkg/networking/core/v1alpha3/gateway.go

```diff
@@ -124,54 +129,75 @@ func (configgen *ConfigGeneratorImpl) buildGatewayListeners(builder *ListenerBui
+		newFilterChains := make([]istionetworking.FilterChain, 0)
 		if p.IsHTTP() {
 			// We have a list of HTTP servers on this port. Build a single listener for the server port.
 			// We only need to look at the first server in the list as the merge logic
 			// ensures that all servers are of same type.
 			port := &networking.Port{Number: port.Number, Protocol: port.Protocol}
 			opts.filterChainOpts = []*filterChainOpts{configgen.createGatewayHTTPFilterChainOpts(builder.node, port, nil, ms.RouteName, proxyConfig)}
-			filterChains = append(filterChains, istionetworking.FilterChain{ListenerProtocol: istionetworking.ListenerProtocolHTTP})
+			newFilterChains = append(newFilterChains, istionetworking.FilterChain{ListenerProtocol: istionetworking.ListenerProtocolHTTP})
 		} else {
 			// build http connection manager with TLS context, for HTTPS servers using simple/mutual TLS
 			// build listener with tcp proxy, with or without TLS context, for TCP servers
 			//   or TLS servers using simple/mutual/passthrough TLS
 			//   or HTTPS servers using passthrough TLS
 			// This process typically yields multiple filter chain matches (with SNI) [if TLS is used]
-			filterChainOpts := make([]*filterChainOpts, 0)
+			newFilterChainOpts := make([]*filterChainOpts, 0)
 
 			for _, server := range servers {
 				if gateway.IsTLSServer(server) && gateway.IsHTTPServer(server) {
 					routeName := mergedGateway.TLSServerInfo[server].RouteName
 					// This is a HTTPS server, where we are doing TLS termination. Build a http connection manager with TLS context
-					filterChainOpts = append(filterChainOpts, configgen.createGatewayHTTPFilterChainOpts(builder.node, server.Port, server,
+					newFilterChainOpts = append(newFilterChainOpts, configgen.createGatewayHTTPFilterChainOpts(builder.node, server.Port, server,
 						routeName, proxyConfig))
-					filterChains = append(filterChains, istionetworking.FilterChain{
+					newFilterChains = append(newFilterChains, istionetworking.FilterChain{
 						ListenerProtocol:   istionetworking.ListenerProtocolHTTP,
 						IstioMutualGateway: server.Tls.Mode == networking.ServerTLSSettings_ISTIO_MUTUAL,
 					})
 				} else {
-					// passthrough or tcp, yields multiple filter chains
-					tcpChainOpts := configgen.createGatewayTCPFilterChainOpts(builder.node, builder.push,
-						server, mergedGateway.GatewayNameForServer[server])
-					filterChainOpts = append(filterChainOpts, tcpChainOpts...)
-					for i := 0; i < len(tcpChainOpts); i++ {
-						filterChains = append(filterChains, istionetworking.FilterChain{ListenerProtocol: istionetworking.ListenerProtocolTCP})
+					// This is the case of TCP or PASSTHROUGH.
+					// Check if we have already built a listener for this - this happens when there are multiple gateways
+					// one with SIMPLE TLS and another with PASSTHROUGH. In that case we should append the filter chains
+					// to the existing listener.
+					mops, exists := mutableopts[lname]
+					if exists {
```

Comment:
This seems a bit odd that only in this spot we are messing with other listeners, and everywhere else we just create a big list of filter chains. I  am a bit concerned this doesn't actually handle all cases. For example, what if we have HTTP and GRPC ports? don't we have the same issue? And why do we have the code here, but not for HTTPS - can't the same happen in that case?

Reply:
Yes. Agree that it is bit odd that is handled in this case. I was n't able to wrap my head around all possibilities. But thinking about it a bit more, we can make this generic. I have changed. I think we will have to add more tests in future.

Reply:
I think te allowed setup are:
* N HTTP servers (megred into a single filtre chain)
* N TLS/HTTPS servers and 0 or 1 TCP server
* 1 TCP server
Is that accurate?

Reply:
Yes. It seems correct.